File: core/src/main/java/org/selyu/smp/core/data/impl/store/MongoProfileStore.java
package org.selyu.smp.core.data.impl.store;

import com.mongodb.client.MongoCollection;
import com.mongodb.client.model.ReplaceOptions;
import org.bson.Document;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.selyu.smp.core.data.store.ProfileStore;
import org.selyu.smp.core.profile.Profile;

import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.regex.Pattern;

import static com.mongodb.client.model.Filters.eq;
import static com.mongodb.client.model.Filters.regex;

public final class MongoProfileStore implements ProfileStore {
    private final MongoCollection<Document> collection;
    private final ReplaceOptions replaceOptions = new ReplaceOptions().upsert(true);

    public MongoProfileStore(@NotNull MongoCollection<Document> collection) {
        this.collection = collection;
    }

    @Override
    public @NotNull CompletableFuture<Optional<Profile>> getByUsername(@NotNull String username) {
        return CompletableFuture.supplyAsync(() -> {
            var document = collection.find(regex("username", Pattern.compile(Pattern.quote(username), Pattern.CASE_INSENSITIVE))).first();
            return Optional.ofNullable(deserialize(document));
        });
    }

    @Override
    public @NotNull CompletableFuture<Optional<Profile>> getByKey(@NotNull UUID uuid) {
        return CompletableFuture.supplyAsync(() -> {
            var document = collection.find(eq("_id", uuid)).first();
            return Optional.ofNullable(deserialize(document));
        });
    }

    @Override
    public @NotNull CompletableFuture<Profile> save(@NotNull Profile profile) {
        return CompletableFuture.supplyAsync(() -> {
            var document = new Document("_id", profile.getUuid());
            document.append("username", profile.getUsername());
            document.append("balance", profile.getBalance());

            collection.replaceOne(eq("_id", profile.getUuid()), document, replaceOptions);
            return profile;
        });
    }

    @Override
    public @NotNull CompletableFuture<Void> delete(@NotNull Profile profile) {
        return CompletableFuture.runAsync(() -> collection.deleteOne(eq("_id", profile.getUuid())));
    }

    @Nullable
    private Profile deserialize(@Nullable Document document) {
        if (document == null)
            return null;

        return new Profile(document.get("_id", UUID.class), document.getString("username"), document.getDouble("balance"));
    }
}

File: core/src/main/kotlin/org/selyu/smp/core/data/mongo/store/MongoProfileStore.kt
package org.selyu.smp.core.data.mongo.store

import com.mongodb.client.MongoCollection
import com.mongodb.client.model.Filters.eq
import com.mongodb.client.model.Filters.regex
import com.mongodb.client.model.ReplaceOptions
import org.bson.Document
import org.selyu.smp.core.data.store.ProfileStore
import org.selyu.smp.core.profile.Profile
import java.util.*
import java.util.concurrent.CompletableFuture
import java.util.concurrent.CompletableFuture.runAsync
import java.util.concurrent.CompletableFuture.supplyAsync
import java.util.regex.Pattern

class MongoProfileStore(private val collection: MongoCollection<Document>) : ProfileStore {
    private val replaceOptions = ReplaceOptions().upsert(true)

    override fun getByUsername(username: String): CompletableFuture<Optional<Profile>> = supplyAsync {
        val document = collection.find(regex("username", Pattern.compile(Pattern.quote(username), Pattern.CASE_INSENSITIVE))).first()
        return@supplyAsync Optional.ofNullable(deserialize(document))
    }

    override fun getByKey(key: UUID): CompletableFuture<Optional<Profile>> = supplyAsync {
        val document = collection.find(eq("_id", key)).first()
        return@supplyAsync Optional.ofNullable(deserialize(document))
    }

    override fun save(value: Profile): CompletableFuture<Profile> = supplyAsync {
        val document = Document("_id", value.uniqueId)
        document.append("username", value.username)
        document.append("balance", value.balance)

        collection.replaceOne(eq("_id", value.uniqueId), document, replaceOptions)
        return@supplyAsync value
    }

    override fun delete(value: Profile): CompletableFuture<Void> = runAsync {
        collection.deleteOne(eq("_id", value.uniqueId))
    }

    private fun deserialize(document: Document?): Profile? {
        if (document == null)
            return null

        return Profile(document.get("_id", UUID::class.java), document.getString("username"), document.getDouble("balance"))
    }
}